File: app/src/main/java/agersant/polaris/features/player/PlayerFragment.java
package agersant.polaris.features.player;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import agersant.polaris.CollectionItem;
import agersant.polaris.PlaybackQueue;
import agersant.polaris.PolarisApplication;
import agersant.polaris.PolarisPlayer;
import agersant.polaris.PolarisState;
import agersant.polaris.R;
import agersant.polaris.api.API;
import agersant.polaris.databinding.FragmentPlayerBinding;
import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;


public class PlayerFragment extends Fragment {

    private boolean seeking = false;
    private BroadcastReceiver receiver;
    private FragmentPlayerBinding binding;
    private ImageView artwork;
    private ImageView pauseToggle;
    private ImageView skipNext;
    private ImageView skipPrevious;
    private SeekBar seekBar;
    private Handler seekBarUpdateHandler;
    private Runnable updateSeekBar;
    private TextView buffering;
    private Toolbar toolbar;
    private API api;
    private PolarisPlayer player;
    private PlaybackQueue playbackQueue;

    private void subscribeToEvents() {
        final PlayerFragment that = this;
        IntentFilter filter = new IntentFilter();
        filter.addAction(PolarisPlayer.PLAYING_TRACK);
        filter.addAction(PolarisPlayer.PAUSED_TRACK);
        filter.addAction(PolarisPlayer.RESUMED_TRACK);
        filter.addAction(PolarisPlayer.COMPLETED_TRACK);
        filter.addAction(PolarisPlayer.OPENING_TRACK);
        filter.addAction(PolarisPlayer.BUFFERING);
        filter.addAction(PolarisPlayer.NOT_BUFFERING);
        filter.addAction(PlaybackQueue.CHANGED_ORDERING);
        filter.addAction(PlaybackQueue.QUEUED_ITEM);
        filter.addAction(PlaybackQueue.QUEUED_ITEMS);
        filter.addAction(PlaybackQueue.REMOVED_ITEM);
        filter.addAction(PlaybackQueue.REMOVED_ITEMS);
        filter.addAction(PlaybackQueue.REORDERED_ITEMS);
        receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                switch (intent.getAction()) {
                    case PolarisPlayer.OPENING_TRACK:
                    case PolarisPlayer.BUFFERING:
                    case PolarisPlayer.NOT_BUFFERING:
                        that.updateBuffering();
                    case PolarisPlayer.PLAYING_TRACK:
                        that.updateContent();
                        that.updateControls();
                        break;
                    case PolarisPlayer.PAUSED_TRACK:
                    case PolarisPlayer.RESUMED_TRACK:
                    case PolarisPlayer.COMPLETED_TRACK:
                    case PlaybackQueue.CHANGED_ORDERING:
                    case PlaybackQueue.REMOVED_ITEM:
                    case PlaybackQueue.REMOVED_ITEMS:
                    case PlaybackQueue.REORDERED_ITEMS:
                    case PlaybackQueue.QUEUED_ITEM:
                    case PlaybackQueue.QUEUED_ITEMS:
                    case PlaybackQueue.OVERWROTE_QUEUE:
                        that.updateControls();
                        break;
                }
            }
        };
        requireActivity().registerReceiver(receiver, filter);
    }

    private void scheduleSeekBarUpdates() {
        updateSeekBar = () -> {
            if (!seeking) {
                int precision = 10000;
                float position = player.getPositionRelative();
                seekBar.setMax(precision);
                seekBar.setProgress((int) (precision * position));
            }
            seekBarUpdateHandler.postDelayed(updateSeekBar, 20/*ms*/);
        };
        seekBarUpdateHandler.post(updateSeekBar);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        setHasOptionsMenu(true);

        PolarisState state = PolarisApplication.getState();
        api = state.api;
        player = state.player;
        playbackQueue = state.playbackQueue;
        seekBarUpdateHandler = new Handler();

        binding = FragmentPlayerBinding.inflate(inflater);
        artwork = binding.artwork;
        pauseToggle = binding.pauseToggle;
        skipNext = binding.skipNext;
        skipPrevious = binding.skipPrevious;
        seekBar = binding.seekBar;
        buffering = binding.buffering;

        toolbar = getActivity().findViewById(R.id.toolbar);

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            int newPosition = 0;

            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                newPosition = progress;
            }

            public void onStartTrackingTouch(SeekBar seekBar) {
                seeking = true;
            }

            public void onStopTrackingTouch(SeekBar seekBar) {
                player.seekToRelative((float) newPosition / seekBar.getMax());
                seeking = false;
                updateControls();
            }
        });

        skipPrevious.setOnClickListener((view) -> player.skipPrevious());
        skipNext.setOnClickListener((view) -> player.skipNext());
        pauseToggle.setOnClickListener((view) -> {
            if (player.isPlaying()) {
                player.pause();
            } else {
                player.resume();
            }
        });

        refresh();

        return binding.getRoot();
    }

    @Override
    public void onStart() {
        subscribeToEvents();
        scheduleSeekBarUpdates();
        super.onStart();
    }

    @Override
    public void onStop() {
        requireActivity().unregisterReceiver(receiver);
        receiver = null;
        super.onStop();
    }

    @Override
    public void onResume() {
        super.onResume();
        refresh();
    }

    private void refresh() {
        updateContent();
        updateControls();
        updateBuffering();
    }

    private void updateContent() {
        CollectionItem currentItem = player.getCurrentItem();
        if (currentItem != null) {
            populateWithTrack(currentItem);
        }
    }

    private void updateControls() {
        final float disabledAlpha = 0.2f;

        int playPauseIcon = player.isPlaying() ? R.drawable.ic_pause_black_24dp : R.drawable.ic_play_arrow_black_24dp;
        pauseToggle.setImageResource(playPauseIcon);
        pauseToggle.setAlpha(player.isIdle() ? disabledAlpha : 1.f);

        if (playbackQueue.hasNextTrack(player.getCurrentItem())) {
            skipNext.setClickable(true);
            skipNext.setAlpha(1.0f);
        } else {
            skipNext.setClickable(false);
            skipNext.setAlpha(disabledAlpha);
        }

        if (playbackQueue.hasPreviousTrack(player.getCurrentItem())) {
            skipPrevious.setClickable(true);
            skipPrevious.setAlpha(1.0f);
        } else {
            skipPrevious.setClickable(false);
            skipPrevious.setAlpha(disabledAlpha);
        }
    }

    private void updateBuffering() {
        if (player.isOpeningSong()) {
            buffering.setText(R.string.player_opening);
        } else if (player.isBuffering()) {
            buffering.setText(R.string.player_buffering);
        }
        if (player.isPlaying() && (player.isOpeningSong() || player.isBuffering())) {
            buffering.setVisibility(View.VISIBLE);
        } else {
            buffering.setVisibility(View.INVISIBLE);
        }
    }

    private void populateWithTrack(CollectionItem item) {
        assert item != null;

        String title = item.getTitle();
        if (title != null) {
            toolbar.setTitle(title);
        }

        String artist = item.getArtist();
        if (artist != null) {
            toolbar.setSubtitle(artist);
        }

        String artworkPath = item.getArtwork();
        if (artworkPath != null) {
            api.loadImageIntoView(item, artwork);
        } else {
            artwork.setImageResource(R.drawable.ic_fallback_artwork);
        }
    }
}

File: app/src/main/java/agersant/polaris/features/player/PlayerFragment.kt
package agersant.polaris.features.player

import agersant.polaris.PlaybackQueue
import agersant.polaris.PolarisApplication
import agersant.polaris.PolarisPlayer
import agersant.polaris.R
import agersant.polaris.api.API
import agersant.polaris.databinding.FragmentPlayerBinding
import agersant.polaris.util.formatTime
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.*
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.google.android.material.progressindicator.CircularProgressIndicator
import com.google.android.material.slider.Slider
import kotlin.math.min
import kotlin.math.roundToInt

class PlayerFragment : Fragment() {
    private val viewModel: PlayerViewModel by viewModels()
    private var seeking = false
    private var receiver: BroadcastReceiver? = null
    private lateinit var artwork: ImageView
    private lateinit var titleText: TextView
    private lateinit var albumText: TextView
    private lateinit var artistText: TextView
    private lateinit var pauseToggle: ImageView
    private lateinit var skipNext: ImageView
    private lateinit var skipPrevious: ImageView
    private lateinit var positionText: TextView
    private lateinit var durationText: TextView
    private lateinit var seekBar: Slider
    private lateinit var buffering: CircularProgressIndicator
    private lateinit var seekBarUpdateHandler: Handler
    private lateinit var updateSeekBar: Runnable
    private lateinit var api: API
    private lateinit var player: PolarisPlayer
    private lateinit var playbackQueue: PlaybackQueue

    private fun subscribeToEvents() {
        val filter = IntentFilter().apply {
            addAction(PolarisPlayer.PLAYING_TRACK)
            addAction(PolarisPlayer.PAUSED_TRACK)
            addAction(PolarisPlayer.RESUMED_TRACK)
            addAction(PolarisPlayer.COMPLETED_TRACK)
            addAction(PolarisPlayer.OPENING_TRACK)
            addAction(PolarisPlayer.BUFFERING)
            addAction(PolarisPlayer.NOT_BUFFERING)
            addAction(PlaybackQueue.CHANGED_ORDERING)
            addAction(PlaybackQueue.QUEUED_ITEM)
            addAction(PlaybackQueue.QUEUED_ITEMS)
            addAction(PlaybackQueue.REMOVED_ITEM)
            addAction(PlaybackQueue.REMOVED_ITEMS)
            addAction(PlaybackQueue.REORDERED_ITEMS)
        }
        receiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                when (intent.action) {
                    PolarisPlayer.OPENING_TRACK,
                    PolarisPlayer.BUFFERING,
                    PolarisPlayer.NOT_BUFFERING -> {
                        this@PlayerFragment.updateContent()
                        this@PlayerFragment.updateControls()
                        this@PlayerFragment.updateBuffering()
                    }
                    PolarisPlayer.PLAYING_TRACK -> {
                        this@PlayerFragment.updateContent()
                        this@PlayerFragment.updateControls()
                    }
                    PolarisPlayer.PAUSED_TRACK,
                    PolarisPlayer.RESUMED_TRACK,
                    PolarisPlayer.COMPLETED_TRACK,
                    PlaybackQueue.CHANGED_ORDERING,
                    PlaybackQueue.REMOVED_ITEM,
                    PlaybackQueue.REMOVED_ITEMS,
                    PlaybackQueue.REORDERED_ITEMS,
                    PlaybackQueue.QUEUED_ITEM,
                    PlaybackQueue.QUEUED_ITEMS,
                    PlaybackQueue.OVERWROTE_QUEUE -> {
                        this@PlayerFragment.updateControls()
                    }
                }
            }
        }
        requireActivity().registerReceiver(receiver, filter)
    }

    private fun scheduleSeekBarUpdates() {
        updateSeekBar = Runnable {
            val duration = player.duration / 1000f
            val position = min(player.currentPosition / 1000f, duration)
            val relativePosition = if (duration != 0f) position / duration else 0f

            if (!seeking) seekBar.value = relativePosition
            durationText.text = formatTime(duration.roundToInt())
            positionText.text = formatTime(position.roundToInt())
            seekBarUpdateHandler.postDelayed(updateSeekBar, 20 /*ms*/)
        }
        seekBarUpdateHandler.post(updateSeekBar)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        setHasOptionsMenu(true)

        val state = PolarisApplication.getState()
        api = state.api
        player = state.player
        playbackQueue = state.playbackQueue
        seekBarUpdateHandler = Handler(Looper.getMainLooper())

        val binding = FragmentPlayerBinding.inflate(inflater)
        artwork = binding.artwork
        titleText = binding.controls.title
        albumText = binding.controls.album
        artistText = binding.controls.artist
        pauseToggle = binding.controls.play
        skipNext = binding.controls.next
        skipPrevious = binding.controls.previous
        positionText = binding.controls.position
        durationText = binding.controls.duration
        seekBar = binding.controls.seekBar
        buffering = binding.controls.buffering

        seekBar.addOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {
                seeking = true
            }

            override fun onStopTrackingTouch(slider: Slider) {
                player.seekToRelative(slider.value / slider.valueTo)
                updateControls()
                seeking = false
            }
        })
        seekBar.setLabelFormatter { value -> formatTime((value * player.duration / 1000f).roundToInt()) }
        skipPrevious.setOnClickListener { player.skipPrevious() }
        skipNext.setOnClickListener { player.skipNext() }
        pauseToggle.setOnClickListener {
            if (player.isPlaying) {
                player.pause()
            } else {
                player.resume()
            }
        }

        if (viewModel.detailsShowing) {
            showDetails()
        }

        return binding.root
    }

    override fun onStart() {
        refresh()
        subscribeToEvents()
        scheduleSeekBarUpdates()
        super.onStart()
    }

    override fun onStop() {
        requireActivity().unregisterReceiver(receiver)
        receiver = null
        seekBarUpdateHandler.removeCallbacks(updateSeekBar)
        super.onStop()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.now_playing, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.show_details -> {
                showDetails()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun refresh() {
        updateContent()
        updateControls()
        updateBuffering()
    }

    private fun updateContent() {
        val item = player.currentItem

        val unknown by lazy { getString(R.string.player_unknown) }
        titleText.text = item?.title ?: unknown
        albumText.text = item?.album ?: unknown
        artistText.text = item?.artist ?: unknown

        if (item?.artwork != null) {
            api.loadImageIntoView(item, artwork)
        } else {
            artwork.setImageResource(R.drawable.ic_fallback_artwork)
        }
    }

    private fun updateControls() {
        val disabledAlpha = 0.2f
        val isNotIdle = !player.isIdle

        if (player.isPlaying && isNotIdle) {
            pauseToggle.setImageResource(R.drawable.ic_round_pause_24)
        } else {
            pauseToggle.setImageResource(R.drawable.ic_round_play_arrow_24)
        }
        pauseToggle.isClickable = isNotIdle
        pauseToggle.alpha = if (isNotIdle) 1f else disabledAlpha

        val hasNextTrack = playbackQueue.hasNextTrack(player.currentItem)
        skipNext.isClickable = hasNextTrack
        skipNext.alpha = if (hasNextTrack) 1f else disabledAlpha

        val hasPrevTrack = playbackQueue.hasPreviousTrack(player.currentItem)
        skipPrevious.isClickable = hasPrevTrack
        skipPrevious.alpha = if (hasPrevTrack) 1f else disabledAlpha

        seekBar.isEnabled = isNotIdle
        positionText.isEnabled = isNotIdle
        durationText.isEnabled = isNotIdle
    }

    private fun updateBuffering() {
        if (player.isOpeningSong || player.isBuffering) {
            buffering.show()
            positionText.visibility = View.INVISIBLE
            durationText.visibility = View.INVISIBLE
        } else {
            buffering.hide()
            positionText.visibility = View.VISIBLE
            durationText.visibility = View.VISIBLE
        }
    }

    private fun showDetails() {
        val item = player.currentItem ?: return

        val dialog = requireContext().showDetailsDialog(item)
        dialog.setOnDismissListener {
            viewModel.detailsShowing = false
        }
        viewModel.detailsShowing = true
    }
}
